### PR TITLE
Demo code for subdivided bilinear grid

### PR DESCRIPTION
**Pull request by akaJes**
draft for additional catmull-rom interpolation of Automatic bed leveling points for bilinear interpolation

### CODE REVIEW

**Comment by thinkyhead** [comment]
@akaJes If we can neaten this up, remove tabs, etc., then I suppose it's a good idea to include it as an experimental feature, get it out there for folks to play with. It's hard to find enough testers, but every release generates a lot of extra interest.

**Comment by akaJes** [comment]
I do not know your style of design code, and if you can help to finish the code, I will be very happy

**Comment by thinkyhead** [comment]
> I do not know your style of design code

Now you do! https://github.com/MarlinFirmware/Marlin/wiki/DNE-Coding-Standards

**Comment by akaJes** [comment]
is this version  more acceptable?

**Comment by thinkyhead** [comment]
I'm just getting around to reviewing this. I'll have some comments soon.

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -8451,6 +8529,18 @@ void ok_to_send() {
 
 #if ENABLED(AUTO_BED_LEVELING_BILINEAR)
 
+  #if ENABLED(ABL_BILINEAR_INTERPOLATE)
+    #define ABL_BG_SPACING  bilinear_grid_spacing_virt
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8533]:
```cpp
#define ABL_BG_SPACING(A)  bilinear_grid_spacing_virt[A]
```

File: Marlin/Marlin_main.cpp [hunk 2]
```diff
@@ -8451,6 +8529,18 @@ void ok_to_send() {
...
+    #define ABL_BG_POINTS_X ABL_GRID_POINTS_VIRT_X
+    #define ABL_BG_POINTS_Y ABL_GRID_POINTS_VIRT_Y
+    #define ABL_BG_GRID     bed_level_grid_virt
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8536]:
```cpp
#define ABL_BG_GRID(X,Y)     bed_level_grid_virt[X][Y]
```

File: Marlin/Marlin_main.cpp [hunk 3]
```diff
@@ -8459,14 +8549,14 @@ void ok_to_send() {
                 y = RAW_Y_POSITION(cartesian[Y_AXIS]) - bilinear_start[Y_AXIS];
 
     // Convert to grid box units
-    float ratio_x = x / bilinear_grid_spacing[X_AXIS],
-          ratio_y = y / bilinear_grid_spacing[Y_AXIS];
+    float ratio_x = x / ABL_BG_SPACING[X_AXIS],
+          ratio_y = y / ABL_BG_SPACING[Y_AXIS];
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8553]:
```cpp
    float ratio_x = x / ABL_BG_SPACING(X_AXIS),
          ratio_y = y / ABL_BG_SPACING(Y_AXIS);
```

File: Marlin/Marlin_main.cpp [hunk 4]
```diff
@@ -8475,10 +8565,10 @@ void ok_to_send() {
...
 
     // Z at the box corners
-    const float z1 = bed_level_grid[gridx][gridy],  // left-front
-                z2 = bed_level_grid[gridx][nexty],  // left-back
-                z3 = bed_level_grid[nextx][gridy],  // right-front
-                z4 = bed_level_grid[nextx][nexty],  // right-back
+    const float z1 = ABL_BG_GRID[gridx][gridy],  // left-front
+                z2 = ABL_BG_GRID[gridx][nexty],  // left-back
+                z3 = ABL_BG_GRID[nextx][gridy],  // right-front
+                z4 = ABL_BG_GRID[nextx][nexty],  // right-back
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8571]:
```cpp
    const float z1 = ABL_BG_GRID(gridx, gridy),  // left-front
                z2 = ABL_BG_GRID(gridx, nexty),  // left-back
                z3 = ABL_BG_GRID(nextx, gridy),  // right-front
                z4 = ABL_BG_GRID(nextx, nexty),  // right-back
```

File: Marlin/Marlin_main.cpp [hunk 5]
```diff
@@ -8826,7 +8916,7 @@ void set_current_from_steppers_for_axis(const AxisEnum axis) {
 
 #elif ENABLED(AUTO_BED_LEVELING_BILINEAR) && !IS_KINEMATIC
 
-  #define CELL_INDEX(A,V) ((RAW_##A##_POSITION(V) - bilinear_start[A##_AXIS]) / bilinear_grid_spacing[A##_AXIS])
+  #define CELL_INDEX(A,V) ((RAW_##A##_POSITION(V) - bilinear_start[A##_AXIS]) / ABL_BG_SPACING[A##_AXIS])
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8919]:
```cpp
#define CELL_INDEX(A,V) ((RAW_##A##_POSITION(V) - bilinear_start[A##_AXIS]) / ABL_BG_SPACING(A##_AXIS))
```

File: Marlin/Marlin_main.cpp [hunk 6]
```diff
@@ -8857,14 +8947,14 @@ void set_current_from_steppers_for_axis(const AxisEnum axis) {
     int8_t gcx = max(cx1, cx2), gcy = max(cy1, cy2);
     if (cx2 != cx1 && TEST(x_splits, gcx)) {
       memcpy(end, destination, sizeof(end));
-      destination[X_AXIS] = LOGICAL_X_POSITION(bilinear_start[X_AXIS] + bilinear_grid_spacing[X_AXIS] * gcx);
+      destination[X_AXIS] = LOGICAL_X_POSITION(bilinear_start[X_AXIS] + ABL_BG_SPACING[X_AXIS] * gcx);
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8950]: `ABL_BG_SPACING(X_AXIS)`

File: Marlin/Marlin_main.cpp [hunk 7]
```diff
@@ -8857,14 +8947,14 @@ void set_current_from_steppers_for_axis(const AxisEnum axis) {
...
       normalized_dist = (destination[X_AXIS] - current_position[X_AXIS]) / (end[X_AXIS] - current_position[X_AXIS]);
       destination[Y_AXIS] = LINE_SEGMENT_END(Y);
       CBI(x_splits, gcx);
     }
     else if (cy2 != cy1 && TEST(y_splits, gcy)) {
       memcpy(end, destination, sizeof(end));
-      destination[Y_AXIS] = LOGICAL_Y_POSITION(bilinear_start[Y_AXIS] + bilinear_grid_spacing[Y_AXIS] * gcy);
+      destination[Y_AXIS] = LOGICAL_Y_POSITION(bilinear_start[Y_AXIS] + ABL_BG_SPACING[Y_AXIS] * gcy);
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8957]: `ABL_BG_SPACING(Y_AXIS)`

**Review by thinkyhead — COMMENTED** [review]
A number of changes are needed. I will just duplicate this PR and submit them, as we're in a minor rush to get RC8 out.

**Comment by thinkyhead** [comment]
Superseded by #5388

**Pull request closed without merging**